### ISSUE TITLE
Completion of empty strings

### ISSUE THREAD

**Issue by gathanase** [elastic/elasticsearch#19147]
This feature request is for the completion suggester, to enable the completion of empty text.
There is currently no way to complete empty text, it returns an empty result list (tested on ES 2.2.1).

Use case scenario:
I need to auto-complete the name of cities given a postal code. Usually only one city matches a postal code, but some postal codes have many cities. For example postal_code='75000' has only one city='Paris', whereas postal_code='51300' has 46 cities.

Of course a prefix query could do the job, but the auto-completion should do the job.

**Comment by clintongormley** [comment]
@areek what do you think?

**Comment by s1monw** [comment]
I am not sure if we should do that... we don't really have suggestions for the empty string, if you as a caller of the API realize that the query is empty you can just run a plain query to get whatever you need? I don't see what we can suggest on the empty query, it's likely something static or heavily user dependent?

**Comment by clintongormley** [comment]
A query wouldn't work given that the candidates are in the completion FST.  We could just provide the first candidates, ranked by whatever weight is specified in the FST.  This is more useful than it would appear at first glance given that the user is using a context which will narrow down the possible candidates already.

**Comment by gathanase** [comment]
Exactly, the context narrows down the candidates.
In my use case, there are 38000 french cities, and usually only 1 (up to 50) for any given postal code.

**Comment by s1monw** [comment]
@areek I am not sure what the cost is when we traverse the FST by weight from the root arcs can you tell if that is feasible?

**Comment by areek** [comment]
One way to accomplish this with completion suggester in v5 is to use a regex prefix (set `regex: "."` instead of setting `prefix`). Maybe we should default to this, when no prefix or regex is specified in the completion query? With this approach, all unique first characters would be queued upon FST intersection and only the highest `5` (default size is 5) paths would be traversed. WDYT?

**Comment by clintongormley** [comment]
I like the idea and, from your description, it sounds pretty efficient.  +1

**Comment by s1monw** [comment]
sounds good to me @areek

**Comment by jimczi** [comment]
cc @elastic/es-search-aggs

**Comment by jtibshirani** [comment]
I started to look into fixing this, but ran into a bug around completion suggestions with regex `"."`. In particular, using the regex prefix in a context suggestion leads to assertion failures in Lucene's `ContextQuery`.

I filed a Lucene issue with more details: https://issues.apache.org/jira/browse/LUCENE-8288. I'll plan to pick this back up once we get some clarity on the issue. Any alternative suggestions are also much appreciated.

**Comment by dubey01** [comment]
Hi guys, this is my first open-source contribution and I found this bug to be quite interesting, I would like to work on this and contribute to this, but as of now, I am a beginner, so can anyone help me out, how should I start on this, and is it possible to get the code base for this project on which I can work Upon.

**Comment by jtibshirani** [comment]
Hello @dubey01, it's great that you're interested in making a contribution. We actually ran into some difficulties addressing this bug, as it involves a lucene issue that is not easy to fix. I removed the 'help wanted' label since it is not very straightforward for a contributor to tackle.

For your first contribution, it would be good to select another issue labeled `help wanted`. We also have the label `good first issue`, which indicates a good issue to tackle for a new contributor: https://github.com/elastic/elasticsearch/labels/good%20first%20issue.
